### PR TITLE
use frozen string literal for ruby 2.3.0+

### PR DESCRIPTION
**Pull request by twalpole**

### CODE REVIEW

File: lib/mail/elements/content_location_element.rb [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
     def location
       @location
     end
-    
+
     def to_s(*args)
       location.to_s
     end
-    
+
```
Comment (jeremy) [inline lib/mail/elements/content_location_element.rb]: Superfluous whitespace?

File: lib/mail/fields/content_disposition_field.rb [hunk 2]
```diff
@@ -9,6 +10,7 @@ class ContentDispositionField < StructuredField
 
     def initialize(value = nil, charset = 'utf-8')
       self.charset = charset
+      value = value.dup if value.is_a?(String) && value.frozen?
       ensure_filename_quoted(value)
       super(CAPITALIZED_FIELD, strip_field(FIELD_NAME, value), charset)
       self.parse
```
Comment (jeremy) [inline lib/mail/fields/content_disposition_field.rb]: Perhaps `#ensure_filename_quoted` shouldn't be modifying `value` in-place, instead?

File: lib/mail/fields/unstructured_field.rb [hunk 3]
```diff
@@ -127,7 +128,7 @@ def fold(prepend = 0) # :nodoc:
           if first
             first = !first
           else
-            word = " " << word
+            word = " #{word}"
```
Comment (jeremy) [inline lib/mail/fields/unstructured_field.rb]: 👍

File: lib/mail/utilities.rb [hunk 4]
```diff
@@ -224,7 +225,7 @@ def map_with_index( enum, &block )
       end
 
     end
-    
+
```
Comment (jeremy) [inline lib/mail/utilities.rb]: Whitespace

File: spec/mail/elements/date_time_element_spec.rb [hunk 5]
```diff
@@ -1,3 +1,5 @@
+# encoding: utf-8
```
Comment (jeremy) [inline spec/mail/elements/date_time_element_spec.rb]: Is the explicit encoding declaration necessary?

File: spec/mail/encodings_spec.rb [hunk 6]
```diff
@@ -837,7 +838,7 @@
   describe ".charset_encoder" do
     class CustomEncoder
       def encode(str, charset)
-        "#{str}-#{charset}"
+        String.new("#{str}-#{charset}")
```
Comment (jeremy) [inline spec/mail/encodings_spec.rb]: Does this need to be a non-frozen string?

File: spec/mail/fields/common/common_field_spec.rb [hunk 7]
```diff
@@ -16,15 +17,15 @@
...
     it "should leave ascii alone" do
       field = Mail::SubjectField.new("This is a test")
       expect(field.encoded).to eq "Subject: This is a test\r\n"
       expect(field.decoded).to eq "This is a test"
     end
-    
+
     it "should encode a utf-8 string as utf-8 quoted printable" do
-      value = "かきくけこ"
+      value = String.new("かきくけこ")
```
Comment (jeremy) [inline spec/mail/fields/common/common_field_spec.rb]: Why does this value need to be mutable?

File: spec/mail/fields/unstructured_field_spec.rb [hunk 8]
```diff
@@ -150,7 +151,7 @@
 
     it "should fold itself if it is non us-ascii" do
       @original = $KCODE if RUBY_VERSION < '1.9'
-      string = "This is あ really long string This is あ really long string This is あ really long string This is あ really long string This is あ really long string"
+      string = String.new("This is あ really long string This is あ really long string This is あ really long string This is あ really long string This is あ really long string")
       @field = Mail::UnstructuredField.new("Subject", string)
       if string.respond_to?(:force_encoding)
         string = string.force_encoding('UTF-8')
```
Comment (jeremy) [inline spec/mail/fields/unstructured_field_spec.rb]: Would `#dup` at this point rather than doing `String.new`

File: spec/mail/fields/unstructured_field_spec.rb [hunk 9]
```diff
@@ -165,7 +166,7 @@
...
+      string = String.new(%|{"unique_args": {"mailing_id":147,"account_id":2}, "to": ["larspind@gmail.com"], "category": "mailing", "filters": {"domainkeys": {"settings": {"domain":1,"enable":1}}}, "sub": {"{{open_image_url}}": ["http://betaling.larspind.local/O/token/147/Mailing::FakeRecipient"], "{{name}}": ["[FIRST NAME]"], "{{signup_reminder}}": ["(her kommer til at stå hvornår folk har skrevet sig op ...)"], "{{unsubscribe_url}}": ["http://betaling.larspind.local/U/token/147/Mailing::FakeRecipient"], "{{email}}": ["larspind@gmail.com"], "{{link:308}}": ["http://betaling.larspind.local/L/308/0/Mailing::FakeRecipient"], "{{confirm_url}}": [""], "{{ref}}": ["[REF]"]}}|)
       @field = Mail::UnstructuredField.new("X-SMTPAPI", string)
       if string.respond_to?(:force_encoding)
         string = string.force_encoding('UTF-8')
```
Comment (jeremy) [inline spec/mail/fields/unstructured_field_spec.rb]: Ditto

File: spec/mail/header_spec.rb [hunk 10]
```diff
@@ -557,13 +558,13 @@
       expect(@traced_header.fields[5].info).to eq "from agw2 by xxx.xxxx.xxx"
       expect(@traced_header.fields[6].field.class).to eq Mail::ToField
     end
-    
+
```
Comment (jeremy) [inline spec/mail/header_spec.rb]: Whitespace

File: spec/mail/utilities_spec.rb [hunk 11]
```diff
@@ -403,7 +404,7 @@
     it "should say an empty string is blank" do
       expect(Mail::Utilities.blank?('')).to be_truthy
       expect(Mail::Utilities.blank?(" ")).to be_truthy
-      a = ''; 1000.times { a << ' ' }
+      a = String.new(''); 1000.times { a << ' ' }
```
Comment (jeremy) [inline spec/mail/utilities_spec.rb]:
```ruby
a = ' ' * 1000
```

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -47,7 +48,7 @@ def fixture(*name)
 # of the letters and numbers work.
 def ascii(from = 33, to = 126)
   chars = []
-  from.upto(to) { |c| chars << ('' << c) }
+  from.upto(to) { |c| chars << (String.new('') << c) }
```
Comment (jeremy) [inline spec/spec_helper.rb]:
```ruby
chars = (from..to).map(&:chr)
```

**Comment by jeremy** [comment]
Nice work, @twalpole! Good call including the pragma everywhere.

Some of these cases suggest that methods are mutating strings that they probably shouldn't.

`String.new(…)` makes a copy of the argument, so selective use of `#dup` may end up feeling cleaner. Having to shed our literal strings in so many places frequently feels like prematurely avoiding frozen strings.

File: lib/mail/elements/content_location_element.rb (hunk 1 above)
Comment (twalpole) [inline lib/mail/elements/content_location_element.rb]: The change removes the superfluous whitespace that was in the file (my editor did that I guess) -- if you're saying the line is unneeded, it was there before but with some spaces at the beginning

File: lib/mail/fields/content_disposition_field.rb (hunk 2 above)
Comment (twalpole) [inline lib/mail/fields/content_disposition_field.rb]: I'm fine with changing that if you're ok with me changing that behavior ?- it is a private method so it's not public API

File: lib/mail/utilities.rb (hunk 4 above)
Comment (twalpole) [inline lib/mail/utilities.rb]: same as the first -- the change actually removed whitespace

File: spec/mail/elements/date_time_element_spec.rb (hunk 5 above)
Comment (twalpole) [inline spec/mail/elements/date_time_element_spec.rb]: no - not sure why that ended up in there -- leftover from debugging something I guess -- will remove

File: spec/mail/encodings_spec.rb (hunk 6 above)
Comment (twalpole) [inline spec/mail/encodings_spec.rb]: yes - unless Ruby19.q_value_decode is changed to not call force_encoding on the result of it

File: spec/mail/fields/common/common_field_spec.rb (hunk 7 above)
Comment (twalpole) [inline spec/mail/fields/common/common_field_spec.rb]: because force_encoding is called on it at line 33

File: spec/mail/fields/unstructured_field_spec.rb (hunk 9 above)
Comment (twalpole) [inline spec/mail/fields/unstructured_field_spec.rb]: Ditto

File: spec/mail/header_spec.rb (hunk 10 above)
Comment (twalpole) [inline spec/mail/header_spec.rb]: same as above

File: spec/mail/utilities_spec.rb (hunk 11 above)
Comment (twalpole) [inline spec/mail/utilities_spec.rb]: ok -- but not sure this change really has anything to do with a PR thats about frozen string literals

File: spec/spec_helper.rb (hunk 12 above)
Comment (twalpole) [inline spec/spec_helper.rb]: ok -- but not sure this change really has anything to do with a PR thats about frozen string literals

File: spec/mail/fields/unstructured_field_spec.rb (hunk 8 above)
Comment (twalpole) [inline spec/mail/fields/unstructured_field_spec.rb]: calling #dup moves it off to the right end of a really long string and immediately clear that it's not a frozen literal - can change if really wanted

File: lib/mail/fields/common/common_field.rb [hunk 13]
```diff
@@ -51,7 +52,9 @@ def strip_field(field_name, value)
     FILENAME_RE = /\b(filename|name)=([^;"\r\n]+\s[^;"\r\n]+)/
     def ensure_filename_quoted(value)
       if value.is_a?(String)
-        value.sub! FILENAME_RE, '\1="\2"'
+        value.sub FILENAME_RE, '\1="\2"'
+      else
+        value
```
Comment (jeremy) [inline lib/mail/fields/common/common_field.rb]: 👍

File: lib/mail/utilities.rb (hunk 4 above)
Comment (jeremy) [inline lib/mail/utilities.rb]: Yep yep. These are OK, but they're unrelated to the PR so they add a bit of noise for review and to the git history, where a `git blame` will show this commit that has nothing to do with the subject code.

File: spec/mail/encoding_spec.rb [hunk 14]
```diff
@@ -115,13 +116,14 @@
...
+      subject = String.new("This is あ string")
       subject.force_encoding('ISO8859-1') if RUBY_VERSION > '1.9'
       mail.subject = subject
       result = mail[:subject].encoded
-      string = "Subject: =?ISO-8859-1?Q?This_is_=E3=81=82_string?=\r\n"
+      string = String.new("Subject: =?ISO-8859-1?Q?This_is_=E3=81=82_string?=\r\n")
       if RUBY_VERSION > '1.9'
         string.force_encoding('ISO8859-1')
+        result = result.dup if result.frozen?
         result.force_encoding('ISO8859-1')
```
Comment (jeremy) [inline spec/mail/encoding_spec.rb]: Having to use `String.new` everywhere is feeling off.

How about using literal string syntax, instead, and duping as needed:
```ruby
  string = string.dup.force_encoding('ISO8859-1')
  result = result.dup.force_encoding('ISO8859-1')
```

Then we can use standard syntax up-front and skip the `#frozen?` checks.

File: spec/mail/encodings_spec.rb (hunk 6 above)
Comment (jeremy) [inline spec/mail/encodings_spec.rb]: 👍

File: spec/mail/fields/common/common_field_spec.rb (hunk 7 above)
Comment (jeremy) [inline spec/mail/fields/common/common_field_spec.rb]: Hmm… we shouldn't need to force-encode to utf-8 since the source encoding is already utf-8, huh?

File: spec/mail/fields/unstructured_field_spec.rb (hunk 8 above)
Comment (jeremy) [inline spec/mail/fields/unstructured_field_spec.rb]: I think we need to work to keep the literal string syntax where we can. We can do `string = string.dup.force_encoding(…)` in many of these cases.

However, this is another case where the `#force_encoding` look unneeded since the source encoding is already utf-8.

File: spec/mail/fields/unstructured_field_spec.rb (hunk 9 above)
Comment (jeremy) [inline spec/mail/fields/unstructured_field_spec.rb]: 😊

File: spec/mail/fields/common/common_field_spec.rb (hunk 7 above)
Comment (twalpole) [inline spec/mail/fields/common/common_field_spec.rb]: I assumed maybe 1.9 didn't obey the encoding or something -- can't remember off the top of my head what versions supported what.  Since it was in the test already I wasn't interested in removing it and if it can be removed it should probably be in another PR

**Comment by twalpole** [comment]
@jeremy Ok -I've swapped to dup in most places and also cleaned up the raise_error warnings to make sure they weren't hiding frozen string errors.

File: spec/mail/fields/common/common_field_spec.rb (hunk 7 above)
Comment (jeremy) [inline spec/mail/fields/common/common_field_spec.rb]: Changing in another PR works too.

Magic comments were added in 1.9.0: https://github.com/ruby/ruby/commit/b1e97c9070e61ce4e7f1be0a02cef9ef302bba8f

File: lib/mail/attachments_list.rb [hunk 15]
```diff
@@ -71,6 +72,7 @@ def []=(name, value)
 
       if hash[:body].respond_to? :force_encoding and hash[:body].respond_to? :valid_encoding?
         if not hash[:body].valid_encoding? and default_values[:content_transfer_encoding].downcase == "binary"
+          hash[:body] = hash[:body].dup if hash[:body].frozen?
           hash[:body].force_encoding("BINARY")
```
Comment (jeremy) [inline lib/mail/attachments_list.rb]: Another spot to sidestep the frozen check:
```ruby
hash[:body] = hash[:body].dup.force_encoding("BINARY")
```

**Comment by jeremy** [comment]
Looking good @twalpole!

Would you squash down to a single commit?

File: lib/mail/attachments_list.rb (hunk 15 above)
Comment (twalpole) [inline lib/mail/attachments_list.rb]: This is in the gems code as opposed to test, and I figured the body could be large -- for performance and memory reasons I didn't want to dup it unless necessary

**Comment by twalpole** [comment]
@jeremy done

**Comment by aried3r** [comment]
Oh, that's great!

I guess that fixes #965 except the method I mentioned there also uses `.force_encoding` but was not modified in this PR. Is the input of the method now already "sanitized"?

**Comment by twalpole** [comment]
@aried3r  The one string `quote_phrase` is called with in the code (display_name) is now mutable so from that perspective it's fixed - If using the method from outside it still forces the encoding on the string and then forces it back to the original encoding - I'll do a quick PR for that
